adding H5MDWriter to H5MD.py

edisj opened the pull request:
Fixes #2866 

Changes made in this Pull Request:
 - adds `class H5MDWriter` to `H5MD.py` 

This is a continuation from #2787 and #2869. I closed #2869 because I tried rebasing to current develop and ended up making a mess of things, so it was easier to make a new pull request.

At the moment, the writer has all of the functionality it needs. It writes all of the data from the `timestep` (positions, velocities, etc), it writes everything from the `ts.data` dictionary that isn't `'time', 'step', 'dt'` into the 'observables' group. It can write datasets with chunked and/or compressed configurations, and files can be opened with MPI drivers.

The current to-do list is:
- ~~Benchmark the writer on multiple nodes with different configurations (chunked, compression)~~
- Have the code reviewed and clean things up
- ~~Figure out what to do the shape argument for `step` and `time` datasets when the number of frames written isn't known by the writer. This could be solved by chunking, but I'm not sure what the default chunk size should be for these datasets.~~

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

pep8speaks commented:
Hello @edisj! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/coordinates/H5MD.py`](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py):

> [Line 1027:50](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1027): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1036:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1036): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 1047:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1047): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> [Line 1048:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1048): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 1049:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1049): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 1050:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1050): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 1064:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1064): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 1065:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1065): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 1066:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1066): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 1079:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1079): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (159 > 79 characters)
> [Line 1080:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1080): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 1081:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1081): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 1083:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1083): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (146 > 79 characters)
> [Line 1084:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1084): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 79 characters)
> [Line 1103:34](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1103): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 1111:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1111): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 1115:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1115): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 1116:80](https://github.com/MDAnalysis/mdanalysis/blob/c064581586192ccb6b8b6e702e481eb88cecf091/package/MDAnalysis/coordinates/H5MD.py#L1116): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

IAlibay commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -826,3 +826,301 @@ def __setstate__(self, state):
+        bool
+            Return ``True`` if `ts` contains a valid simulation box
+        """
+        return np.all(ts.dimensions > 0)
```
Just passing by (strangely enough was thinking of #2698 and saw this pop up), going by line 622, wouldn't a no-dimensions case set `ts.dimensions` to `None` (i.e. this should cause a `TypeError`)?

IAlibay commented:
@edisj @orbeckst what's the target milestone for this? (i.e. is this a long term goal or are you planning 2.0/2.1?)

orbeckst commented:
It's not crucial for 2.0 and could go into 2.1 (at this point there will likely be a citation for a SciPy paper); if @edisj wants to work double-time to attempt to squeeze it into a 2.0 then I won't object, though.

edisj commented:
Hi @IAlibay , I'll push to have it done by 2.0 (May 10th, right?). I'm still testing some things with HPC benchmarks before updating this PR

IAlibay commented:
> Hi @IAlibay , I'll push to have it done by 2.0 (May 10th, right?).

So unless I misunderstood our workshop targets, the 10th is really the day it needs to be on conda-forge. I think we'd have to put down the code freeze at the latest the Friday before, i.e. the May 7th.

edisj commented:
Hey @orbeckst , ready for review! :) (also @IAlibay if  wanted to take a look)

I've made a lot of changes over the last few days and think it's complete other than  maybe a couple exceptions to add and documentation to edit. I'll finish up the tests for codecov tomorrow

There are a few parts of the code I'm not sure about, I'll open some discussions...

IAlibay reviewed (changes requested):
Sorry only had time to have a quick glance over, so it's mostly superficial comments.

I'll probably wait until you've added the extra tests before I re-review.

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+                             driver='wrong_driver', comm="MPI.COMM_WORLD")
+
+    def test_open_with_driver(self):
+        u = mda.Universe(TPR_xvf, H5MD_xvf, driver="core")
```
`assert` something here

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+    @pytest.mark.xfail(reason='Issue #2884')
+    def test_open_filestream(self, h5md_file):
+        with h5md_file as f:
+            u = mda.Universe(TPR_xvf, h5md_file)
```
`assert` something here

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
                 del g['particles/trajectory/box/edges/value'].attrs['unit']
         with pytest.raises(ValueError):
-            u = mda.Universe(TPR_xvf, outfile, format="H5MD")
+            u = mda.Universe(TPR_xvf, outfile)
```
`assert` something here

edisj replied:
should this be asserting something? I thought it was just trying to catch the value error

orbeckst replied:
You're right, I must have overlooked the raises.

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -329,84 +280,28 @@ def test_unknown_time_unit(h5md_file, ref, tmpdir):
-
+                del g['particles/trajectory/position']
+        u = mda.Universe(TPR_xvf, outfile)
+        assert u.trajectory.units['length'] == 'nm'
```
Also check that values are correct?

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
 @pytest.mark.skipif(not HAS_H5PY, reason="h5py not installed")
-def test_open_with_driver():
-    u = mda.Universe(TPR_xvf, H5MD_xvf, driver="core")
+class TestH5MDWriter_2(object):
```
more specific naming, please

orbeckst replied:
and why can't it be under `TestH5MDWriter` ? ;-)

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+            (True, True, True),
+            (False, True, True),
+            (False, False, True),
+            #(False, False, False)
```
What is the behavior in this case? It should be well-defined and tested.

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+        for u1, u2 in zip(units, uw_units):
+            assert_equal(u1, u2)
+
+    #@pytest.mark.parametrize('chunks', (None, (1, 1000, 3)))
```
What does `chunks = None` do at the moment? Trying different chunks looks like a good test.

edisj replied:
new updates let you write contiguous dsets, made a new test for this case

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+
+        uw = mda.Universe(TPR_xvf, outfile)
+        dset = uw.trajectory._particle_group['position/value']
+        assert_equal(dset.chunks, chunks)
```
Should you also assert actual values to make sure that they end up in the correct position in the file?

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+        assert_equal(dset.compression, filter)
+        assert_equal(dset.compression_opts, opts)
+
+    @pytest.mark.parametrize('driver, comm', (('core', None),
+                                              ('mpio', 'comm')))
```
The try/except inside to distinguish these two cases looks unnecessary complicated and introduces additional uncertainty (the blanket `except`). Why not make two different tests?

edisj replied:
two separate tests makes more sense

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+            uw = mda.Universe(TPR_xvf, outfile, driver=driver, comm=comm)
+            file = uw.trajectory._file
+            assert_equal(file.driver, driver)
+        except:
```
That's not good — avoid catch-all excepts as they can obscure real problems.

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+        assert_equal(dset.compression, filter)
+        assert_equal(dset.compression_opts, opts)
+
+    @pytest.mark.parametrize('driver, comm', (('core', None),
```
I assume that 'comm' on the next line is a mock-up for the MPI communicator. If so, perhaps either use a proper `mock` object or write the string like "mock MPICOMM", just to make sure that this is clear. Sometimes people look at the tests for how to use code so when you don't use the code correctly (on purpose) it can get confusing.

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+                    for ts in universe.trajectory:
+                        W.write(universe)
+
+    def test_timestep_not_modified_by_writer(self, universe, Writer, outfile):
```
Is this a concern?''

I am generally all in favor of testing everything but if there's no possibility that this happens, I'd remove it to make the tests pass a shred faster.

orbeckst replied:
As discussed in our meeting, leave it as it seems to be useful and done for other writers.

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -493,9 +358,325 @@ def test_no_units(h5md_file, ref, tmpdir):
+        assert_equal(
+            ts.time, time, err_msg="Time in Timestep was modified by writer.")
+
+    """def test_gaps(self, universe, Writer, outfile):
```
You can mark tests as XFAIL if that's a known problem. It serves as documentation and a good way to motivate future work — all one has to do is to make the XFAILS pass!

orbeckst commented on testsuite/MDAnalysisTests/data/coordinates/README.md:
```diff
@@ -44,10 +44,7 @@ plain text using Gromacs utilities.
-Converted test.trr to test.h5md with MDAnalysis and pyh5md
-using the 'create_h5md_data.py' script.
-See https://github.com/pdebuyl/pyh5md for pyh5md repository.
+Written with MDAnalysis using the 'create_h5md_data.py' script
```
Does the h5md file now have good layout?

edisj replied:
yes, it's following the (1, n_atoms, 3) chunk scheme

orbeckst commented on testsuite/MDAnalysisTests/data/coordinates/README.md:
```diff
@@ -44,10 +44,7 @@ plain text using Gromacs utilities.
 ## Validation
-Manually examined contents of test.trr with MDAnalysis and test.h5md
-with h5py and checked that the contents were the same.
+Manually examined contents of test.h5md with MDAnalysis.
```
Validation should use tools outside MDAnalysis.... otherwise we might be self consistent but wrong.

orbeckst reviewed (changes requested):
That's a pretty hefty PR. Sorry, it took me a while and there's probably stuff I overlooked but please see comments/questions.

Apart from my comments, please update CHANGELOG.

edisj commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -183,10 +184,13 @@
 import MDAnalysis as mda
+import numpy as np
 from . import base, core
 from .base import Timestep
```
Wow I didn't realize I had to hit "submit review" for my comments to show up, I thought I had already replied to your review comments sorry :(

edisj commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +746,433 @@ def __setstate__(self, state):
+                                        " For more information on units, see"
+                                        " `MDAnalyis units`_.")
+
+        """if self.convert_units:
```
Will uncomment and add tests

edisj commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +746,451 @@ def __setstate__(self, state):
+                                       f" {self._unit_translation_dict.keys()}"
+                                        " For more information on units, see"
+                                        " `MDAnalyis units`_.")
+                else:
+                    self.units[key] = self.new_units[key]
```
I'm pretty sure `self.units` has to be in MDA units since all of the `convert_*_to_native()` methods pull from `self.units.values()` to get conversion factors

edisj commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +746,451 @@ def __setstate__(self, state):
+        # step must exist in h5md file
+        except(KeyError):
+            self._step[-1] = ts.frame
+        self._time.resize(self._time.shape[0]+1, axis=0)
```
I've done a little bit of digging but can't seem to find any case where resizing with a particular chunk layout doesn't work. I'll keep looking though

orbeckst replied:
Add a comment to the code here along the lines of "should work with ANY chunk layout", at least then we have a record that we thought about it. If you find any evidence, put links/citation in the comment.

edisj commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +746,451 @@ def __setstate__(self, state):
+                self._step = self.traj.require_dataset(f'{group}/step',
+                                                       shape=(0,),
+                                                       maxshape=(None,),
+                                                       dtype=np.int32)
```
From what I can tell MDA uses the regular python int for both `ts.data['step']` and `ts.frame`. Is there any datatype conversion between int and numpy.int32?

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +746,433 @@ def __setstate__(self, state):
+        self.h5md_file = None
+
+        # check which datasets are to be written
+        self.has_positions = kwargs.get('positions', False)
```
Yes, that's a good idea. We still need the catch-all `**kwargs` for anything else that we do not care about but having explicit kwarg names in the signature (+ defaults) is good.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+        [:attr:`MDAnalysis.__version__`]
+
+
+    Note
+    ----
+    Parellel writing with the use of a MPI communicator and the ``'mpio'``
+    HDF5 driver is currently not supported.
```
Move to after the synopsis/before Parameters. If it doesn't format as a box, use
```reST
.. note::

    Parallel writing ...
```

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+        when any of the optional ``timeunit``, ``lengthunit``,
+        ``velocityunit``, or ``forceunit`` keyword arguments are
+        unrecognizable by MDAnalysis
+
```
Put the long text section here. See numpy doc.

orbeckst replied:
You have a lot of great docs hidden in private methods. Many describe limitations or caveats with the format or writer. The notes section should include this information, too.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+                 chunks=None,
+                 compression=None,
+                 compression_opts=None,
+                 **kwargs):
```
Make all documented kwargs explicit instead of `**kwargs`. Just leave `**kwargs` to catch any unprocessed kwargs.

This will help a lot with the documentation and seeing defaults.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+    def _determine_units(self, ag):
+        """determine which units the file will be written with
+
+        By default, it fills the :attr:`self.units` dictionary by copying the
+        units dictionary of the parent file. Because H5MD files have no
+        standard unit restrictions, users may pass a kwarg in ``(timeunit,
+        lengthunit, velocityunit, forceunit)`` to the writer so long as
+        MDAnalysis has a conversion factor for it (:exc:`ValueError` raised if
+        it does not). These custom unit arguments must be in
+        `MDAnalysis notation`_. If custom units are supplied from the user,
+        :attr`self.units[unit]` is replaced with the corresponding
+        ``unit`` argument.
```
This text will not be visible in the docs even though it contains useful information. Please summarize it as part of the section on units in the writer notes.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+        """
+
+        # use units of parent file as native units
+        try:
+            # ag == AtomGroup
+            self.units = ag.universe.trajectory.units.copy()
+        except AttributeError:
+            # ag == Universe
+            self.units = ag.trajectory.units.copy()
+
```
Replace this whole thing with
```python
self.units = ag.universe.trajectory.units.copy()
```

Should always work (tested in 2.0.0b):
```python
In [1]: import MDAnalysis as mda
from
In [2]: from MDAnalysis.tests import datafiles as data

In [4]: u = mda.Universe(data.TPR, data.XTC)

In [5]: u.universe.trajectory.units
Out[5]: {'time': 'ps', 'length': 'nm'}

In [6]: u.atoms.universe.trajectory.units
Out[6]: {'time': 'ps', 'length': 'nm'}
```

orbeckst replied:
And it will increase coverage because you don't have to explicitly test the two cases.

(We'll assume that it works because hopefully we're testing elsewhere writing with ag and Universe.)

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+        for key, value in self._new_units.items():
+            if value is not None:
+                if value not in self._unit_translation_dict[key]:
+                    raise ValueError(f"{value} is not a unit recognizable by"
```
recognized

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+        """
+
+        if self._comm is not None:  # pragma: no cover
+            # ValueError raised if communicator is passed
```
say where he ValueError is raised, currently a bit confusing because the raise is not here

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+        """initializes all datasets that will be written to by
+        :meth:`_write_next_timestep`
+
+        Note
+        ----
+        :exc:`NoDataError` is raised if no positions, velocities, or forces are
+        found in the input trajectory. While the H5MD standard allows for this
+        case, :class:`H5MDReader` cannot currently read files without at least
+        one of these three groups. A future change to both the reader and
+        writer will allow this case.
```
Does not show up in docs. Summarize in main writer docs (and leave it here, too).

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+        # for keeping track of where to write in the dataset
+        self._counter = 0
+
+        # ask the parent file if it has postions, velocities, and forces
```
positions

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+            self._set_attr_unit(self._edges, 'length')
+            self._set_attr_unit(self._time, 'time')
+        else:
+            self._traj['box'].attrs['boundary'] = 3*['none']
```
comment that it _must_ be "none" and not `None` according to ... (h5 or h5md?)

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+    def _set_attr_unit(self, dset, unit):
+        """helper function to set a 'unit' attribute for an HDF5 dataset"""
+
+        if self.units[unit] is None:
```
Is there really no test case that exercises this branch?

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +777,635 @@ def __setstate__(self, state):
+        The first dimension of each dataset is extended by +1 and
+        then the data is written to the new slot.
+
+        Note
+        ----
+        Writing H5MD files with fancy trajectory slicing where the Timestep
+        does not increase monotonically such as ``u.trajectory[[0,1,2,1]]``
+        or ``u.trajectory[[0,1,2,0,1,2]]`` raises a :exc:`ValueError` as this
+        violates the rules of the step dataset in the H5MD standard.
```
Add this information to the main docs (as this is not visible).

orbeckst reviewed (changes requested):
First half of a review (did not have enough time to finish). Mainly docs.

edisj commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -148,6 +175,7 @@
 .. _`Build h5py from sources`: https://docs.h5py.org/en/stable/mpi.html#building-against-parallel-hdf5
 .. _`H5MD notation`: https://nongnu.org/h5md/modules/units.html
 .. _`MDAnalysis notation`: https://userguide.mdanalysis.org/units.html
+.. _`MDAnalyis units`: https://userguide.mdanalysis.org/units.html
```
fixed

orbeckst commented:
@edisj see the remaining comments, please. If you're addressing it then just resolve it. If you want to discuss or disagree then add a comment and leave it open.

It would also be nice to use *duecredit* to automatically generate citations.

edisj commented:
Hi @orbeckst , just pushed the changes to duecredit and added that final test. All comments should be addressed! I'm going through to add a couple more details in the docs regarding writing files with `n_frames` and with contiguous layouts.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -1002,13 +1000,19 @@ class H5MDWriter(base.WriterBase):
                  velocityunit=None, forceunit=None, author='N/A',
                  author_email=None, creator='MDAnalysis',
-                 creator_version=mda.__version__, **kwargs):
+                 creator_version=mda_version, **kwargs):
```
`mda.__version__` looks ok to me, why did you change it?

edisj replied:
instead of importing all of `mda`, I did: `from MDAnalysis import __version__ as mda_version`, since the only thing I needed to import `mda` for was `mda.__version__`. Is one way preferable over the other?

orbeckst replied:
You will immediately import all of mda with `from MDAnalysis import __version__` — it makes no difference. Therefore, I'd keep it simple and clear (as it was previously). Generally, I am not a fan of `from x import y` because it is messy. "Namespaces are a honking good idea. Let's have more of those" is one of the Python Zens... IIRC.

orbeckst replied:
Still to do

orbeckst commented on package/CHANGELOG:
```diff
@@ -23,9 +23,13 @@ The rules for this file:
   * 2.0.0
 
 Fixes
+<<<<<<< HEAD
+  * Fixed missing array flatten preventing PCA from running when mean positions
+=======
   * Fixed sometimes wrong sorting of atoms into fragments when unwrapping
     (Issue #3352, PR #3353)
   * Fixed missing array flatten preventing PCA from running when mean positions  
+>>>>>>> 65319046c49f4bce5e49949b7ab451039adcbd67
     provided and bugs causing tests to incorrectly pass (Issue #2728, PR #3296)
```
fix merge !!

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_writer_api.py:
```diff
@@ -52,6 +52,9 @@ def test_ts_error(writer, tmpdir):
         fn = str(tmpdir.join('out.xtc'))
     elif writer == mda.coordinates.LAMMPS.DATAWriter:
         pytest.skip("DATAWriter requires integer atom types")
+    elif writer == mda.coordinates.H5MD.H5MDWriter:
+        pytest.skip("H5MDWriter requires h5py to be installed, "
+                    "fails MIN dependency test")
```
You are _always_ excluding this test, even if hdf5 is installed. The API tests should run in full tests.

Only skip the tests if you cannot import hdf5.

orbeckst reviewed (changes requested):
initial comments to change, sorry short on time

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_writer_api.py:
```diff
@@ -52,10 +52,9 @@ def test_ts_error(writer, tmpdir):
-            pytest.skip("H5MDWriter requires h5py to be installed, "
-                        "fails MIN dependency test")
+    elif writer == mda.coordinates.H5MD.H5MDWriter and not mda.coordinates.H5MD.HAS_H5PY:
+        pytest.skip("H5MDWriter requires h5py to be installed, "
+                    "fails MIN dependency test")
```
It's not a fail, just a skip. Just say "skipping for MIN dependency test"

orbeckst replied:
Also, remove 3.6, this might change in the future. ... and the ubuntu/macOS
```suggestion
        pytest.skip("skipping H5MDWriter test because h5py is not installed")
```

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_writer_api.py:
```diff
@@ -83,14 +82,17 @@ def test_write_with_atomgroup(writer, tmpdir):
-        if not mda.coordinates.H5MD.HAS_H5PY:
-            pytest.skip("macOS and ubuntu, 3.6, MIN tests fail")
+    elif writer == mda.coordinates.H5MD.H5MDWriter and not mda.coordinates.H5MD.HAS_H5PY:
+        pytest.skip("macOS and ubuntu, 3.6, MIN tests fail")
```
It's not a fail, just a skip. Just say "skipping for MIN dependency test"

orbeckst replied:
Also, remove 3.6, this might change in the future. ... and the ubuntu/macOS
```suggestion
        pytest.skip("skipping H5MDWriter test because h5py is not installed")
```

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_writer_api.py:
```diff
@@ -83,14 +82,17 @@ def test_write_with_atomgroup(writer, tmpdir):
-    with writer(fn, n_atoms=u.atoms.n_atoms) as w:
-        w.write(u.atoms)
+    if writer == mda.coordinates.H5MD.H5MDWriter:
+        with writer(fn, n_atoms=u.atoms.n_atoms, convert_units=False) as w:
```
add a comment explaining why special case necessary here

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_writer_api.py:
```diff
@@ -106,11 +108,14 @@ def test_write_with_universe(writer, tmpdir):
-        if not mda.coordinates.H5MD.HAS_H5PY:
-            pytest.skip("macOS and ubuntu, 3.6, MIN tests fail")
+    elif writer == mda.coordinates.H5MD.H5MDWriter and not mda.coordinates.H5MD.HAS_H5PY:
+        pytest.skip("macOS and ubuntu, 3.6, MIN tests fail")
```
It's not a fail, just a skip. Just say "skipping for MIN dependency test"

orbeckst replied:
Also, remove 3.6, this might change in the future. ... and the ubuntu/macOS
```suggestion
        pytest.skip("skipping H5MDWriter test because h5py is not installed")
```

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_writer_api.py:
```diff
@@ -106,11 +108,14 @@ def test_write_with_universe(writer, tmpdir):
-    with writer(fn, n_atoms=10) as w:
-        w.write(u)
+    if writer == mda.coordinates.H5MD.H5MDWriter:
+        with writer(fn, n_atoms=u.atoms.n_atoms, convert_units=False) as w:
```
please explain why special case needed (comment)

orbeckst reviewed (changes requested):
minor doc/comment changes

orbeckst commented:
@edisj is there anything missing as far as you can see? Is it worth's @IAlibay 's time to have a final look?

I only had one last set of minor requests, but I also have seen the whole thing multiple times so a second pair of eyes would be welcome.

Can you think of anyone else who should review the hdf5 code, @edisj ?

edisj commented:
> @edisj is there anything missing as far as you can see? Is it worth's @IAlibay 's time to have a final look?
> 
> I only had one last set of minor requests, but I also have seen the whole thing multiple times so a second pair of eyes would be welcome.
> 
> Can you think of anyone else who should review the hdf5 code, @edisj ?

After scanning through the pull request again I see I never updated `testsuite/MDAnalysisTests/data/coordinates/README.md` to say it was validated outside MDAnalysis with h5py, I'll fix that...

Maybe @lilyminium or @IAlibay can scan through it? In the dev discord a while ago, we discussed what the default data to write from the `ts` should be. If you two have some spare time to scan through to make sure it's consistent with the other writers I would appreciate it. Also, what are your thoughts on the documentation? Is it understandable? And is the logic of the design decisions clear?

IAlibay commented:
> Maybe @lilyminium or @IAlibay can scan through it?
Sure I'll have a look through tomorrow morning.

P.S. Self-assigned because it's the only way to fool my notification filters into thinking it has work to do 🙄

IAlibay commented on testsuite/MDAnalysisTests/data/coordinates/create_h5md_data.py:
```diff
@@ -1,81 +1,41 @@
-            data_lambda.append(uni.trajectory.ts.data['lambda'],
-                               ts.frame, time=ts.time)
 
 def main():
```
How is this being called? It's unclear to me as to how this is being used. i.e. do we need to have a `main` or is it just being called by something else?

edisj replied:
It's meant to be called by a person, not within the package itself. I copied `testsuite/MDAnalysisTests/data/coordinates/create_data.py` which does the same thing to create synthetic trajectories

edisj replied:
Unless I'm misunderstanding how `testsuite/MDAnalysisTests/data/coordinates/create_data.py` is meant to be used...

IAlibay replied:
I'll be honest, that one exists from before my time so I don't know either 😅 I'm happy with it as-is if there's a use case for it.

IAlibay commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+                 creator_version=mda_version, **kwargs):
+
+        if not HAS_H5PY:
+            raise RuntimeError("H5MDWriter: Please install h5py")
```
We have the machinery to mock the inability to import of a module. Can we do this to test this?

orbeckst replied:
Check out PR #3294 where @IAlibay mocks out netcdf.

IAlibay commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+                self._force[i] = self.convert_forces_to_native(self._force[i])
+
+    @property
+    def has_positions(self):
```
any ideas why the setters aren't getting covered here?

orbeckst replied:
They are never set through the setter in this code.

Will need extra tests.

IAlibay reviewed (changes requested):
Having a deeper look but here are some initial comments from a quick scan.

IAlibay commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+
+        """
+
+        if self._comm is not None:  # pragma: no cover
```
If this is currently an unreachable code path, would just having this commented out instead be better? (it's not cleaner, but then at least there's no unaccessible code?

edisj replied:
I opted to remove it completely until parallel is enabled

IAlibay commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+        elif self._driver is not None:
+            self.h5md_file = h5py.File(name=self.filename,
+                                       mode='w',
+                                       driver=self._driver)
```
The default value for `driver` is `None` anyways, could this not be squashed into the one call?

edisj replied:
it can, looks much better!

IAlibay commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+        # box group is required for every group in 'particles'
+        self._traj.require_group('box')
+        self._traj['box'].attrs['dimension'] = 3
+        if ts.dimensions is not None and np.all(ts.dimensions > 0):
```
Just checking, is the ts.dimensions == 0 as an empty box behaviour also being replicated on read?

edisj replied:
good call, I added a parameter in the "no dimensions" test and it passes

IAlibay reviewed (changes requested):
--> review through to line 1240 of H5MD.py

Sorry I'm going to have to break up my reviews into chunks. Here is the first part of my review, I'll follow up with the rest probably tomorrow at this rate.

IAlibay commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+
+        """
+
+        i = self._counter
```
A more descriptive counter here might be worth it. I'm not in the "i,j,k,l are terrible" camp, but at the same time if they get created outside of a loop it's really easy for someone to accidentally do `for i in range(N)` somewhere in your code and forget that you had `i` already declared somewhere else.

orbeckst replied:
I think `i` is fine here because it makes the rest of the code more readable.

edisj replied:
I like `i` since it's always used as a counter, but I see your point in that since _it's always used as a counter_ someone might accidently overwrite it. I'll leave it for now?

IAlibay commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -77,336 +75,226 @@ def test_get_writer_2(self, ref, reader, tmpdir):
-        u = mda.Universe(TPR_xvf, outfile, format='H5MD')
+                del g[f'particles/trajectory/{group}']
+        u = mda.Universe(TPR_xvf, outfile)
         with pytest.raises(NoDataError):
```
Wherever possible, can we have a `match` for the associated error/warning we are matching?
This makes sure that we are testing the error we actually want to test (you'd think this wouldn't be needed, but in practice it happens more often than you'd think).

edisj replied:
I'll go through and add matches

IAlibay reviewed (changes requested):
Second half of review.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -613,13 +634,19 @@ def _copy_to_data(self):
         for name, value in self._has.items():
             if value:
                 if 'time' in self._particle_group[name]:
                     self.ts.data['time'] = self._particle_group[name][
```
`ts.data` is actually where per-step information lives... and `ts.time = ...` goes through the property.setter. But I agree, `ts.time = ...` is clearer and more uniform across the code base.

edisj replied:
changed to `ts.time` instead of `ts.data['time']`

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -613,13 +634,19 @@ def _copy_to_data(self):
                     self.ts.data['time'] = self._particle_group[name][
                         'time'][self._frame]
                     break
+        for name, value in self._has.items():
```
I think, technically speaking, `time` and `step` can be in different groups so the `break` for one would not be correct for the other. --- or did I get this wrong @edisj ?

IAlibay replied:
I think I was thinking more of a `continue` statement, but I can see how this might be a pain so I'm happy with the way it is.

edisj replied:
That's right, `time` is not a required dataset, so some groups may contain it and some may not

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+    :exc:`NoDataError` is raised if no positions, velocities, or forces are
+    found in the input trajectory. While the H5MD standard allows for this
+    case, :class:`H5MDReader` cannot currently read files without at least
+    one of these three groups. A future change to both the reader and
```
Maybe do not promise "a future change" and just acknowledge the limitation. Raise an issue and wait for anyone actually wanting empty h5md trajectories (eg for storing observables).

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+    comm : :class:`MPI.Comm` (optional)
+        MPI communicator used to open H5MD file
+        Must be passed with `'mpio'` file driver.
+        This argument is currently disabled.
```
On reflection I agree with @IAlibay .  Keep it lean and promise as little as possible. 

Remove the docs for `comm` and the code that raises the exception. You can raise an issue and copy the text/code there for future use.

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+            raise ValueError("H5MDWriter: no atoms in output trajectory")
+        self._driver = driver
+        self._comm = comm
+        if self._comm is not None or self._driver == 'mpio':
```
remove (with the parallel promise)

orbeckst commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+
+        self._counter += 1
+
+    def _convert_units(self, i):
```
Can you rename this to `_convert_dataset_with_units()` or something similarly expressive? Because for a second I though this was one of the standard `_convert_*` methods and wondered why you would want to convert the integer `i`.

orbeckst reviewed (changes requested):
Minor things & mostly agreeing with @IAlibay (except that I think `i` can stay ;-) ).

You'll have to ask @IAlibay how fast you need to be in order to try and slip this into 2.0.0.

edisj commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -687,6 +714,25 @@ def _reopen(self):
+        return H5MDWriter(
+            filename,
+            n_atoms=n_atoms,
+            convert_units=self.convert_units,
```
I went ahead and mimicked the way you handled kwargs in NCDF. By default it now writes with the settings of the file being read and users should be able to overwrite them

edisj commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -687,6 +714,25 @@ def _reopen(self):
+        """
+        if n_atoms is None:
+            n_atoms = self.n_atoms
+        return H5MDWriter(
```
I agree, fixed now

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -1,13 +1,14 @@
 import pytest
 from numpy.testing import assert_almost_equal, assert_equal
 import numpy as np
+import sys
```
Not needed, I think.
```suggestion
```

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -457,13 +457,23 @@ def test_step_from_frame(self, universe, Writer, outfile):
+            W.write(universe)
+
+        uw = mda.Universe(TPR_xvf, outfile)
+        assert_equal(uw.trajectory.ts.has_positions, True)
+        assert_equal(uw.trajectory.ts.has_velocities, False)
+        assert_equal(uw.trajectory.ts.has_forces, False)
```
Instead of `assert_equal` (which works for everything but particularly arrays) I would just use the simpl
```python
assert uw.trajectory.ts.has_positions is True
```
(for boolean, `is` is ok, but in general `==` is safer).

lilyminium replied:
Interesting facts:

```python
>>> np.bool(1) is True
True
>>> np.ones(1, dtype=bool)[0] is True
False
>>> np.ones(1, dtype=bool)[0] == True
True
```

;)

orbeckst replied:
Scary.

And odd:

```python
>>> isinstance(np.bool_, bool)
False
```

lilyminium replied:
Yeah the basic gist is that `np.bool_` is an actual type, as opposed to the Python bool which is a subclass of int. [There's a warning on the docs if you actively look for it.](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.bool_). It's just funny when `assert True is True` failing.

```python
>>> isinstance(True, int)
True
```

<img width="504" alt="Screen Shot 2021-08-01 at 1 49 18 pm" src="https://user-images.githubusercontent.com/31115101/130301557-5fe0a5d0-2c79-4d90-92dc-ef07771368ad.png">

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_h5md.py:
```diff
@@ -457,13 +457,23 @@ def test_step_from_frame(self, universe, Writer, outfile):
             W.write(universe)
             assert_equal(W.has_positions, W._has['position'])
             assert_equal(W.has_velocities, W._has['velocity'])
             assert_equal(W.has_forces, W._has['force'])
```
See other comment
```python
assert W.has_positions == W._has['position']
```

IAlibay commented:
I can try to re-review a bit later today, although if you think all my comments were addressed please do go ahead with the merge @orbeckst

orbeckst commented:
Your point about "check exception messages" was a good one — I am currently watching @edisj find out which exceptions are actually triggered .... so _my_ comments had been addressed, yours should be done in the next hour.

edisj commented:
Hi @IAlibay , I just pushed my final changes, all comments should be addressed. All error tests now have a `match`. I hope there's still time to make it into 2.0.0...

If you have time to review, there are a couple of decisions made while talking to @orbeckst that might need a little justification:
- Removed the `has_*` setters from the writer because we realized that in order to actually write a test to activate them, we had to write code that wouldn't ever be used for an H5MD trajectory, so the setter itself was pointless
- Passed defaults (compression, compression_opts, driver) from input H5MD into `H5MDReader.Writer()`, but intentionally left out `chunks`. This is because H5MDWriter chooses `(1, n_atoms, 3)` by default, and accidently choosing a bad chunk shape can be really bad for performance. I added a detailed note in the docstring of `H5MDReader.Writer()` to explain the situation and also say how to manually set the chunk shape if you wanted to.

Thanks again for your review :D, if there's anything else that remains let me know

IAlibay reviewed (commented):
Thanks @edisj I'm just having a quick look through now. Just doing this quick pre-review comment to include this duecredit thing - not sure if it needs testing or not ?

IAlibay commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -352,6 +377,12 @@ class H5MDReader(base.ReaderBase):
         }
     }
 
+    @due.dcite(Doi("10.25080/majora-1b6fd038-005"),
```
Or I guess my comment didn't get added to my review - was wondering if this needed to be added to test_duecredit in some way?

edisj replied:
I did test it manually with a script that used H5MD and it works. I asked @orbeckst whether a unit test was necessary and he said no

IAlibay replied:
Ok, sounds fair, wouldn't have thought it would be a blocking thing either way anyways. @orbeckst do we maybe want to raise an issue for the future to check if we should eventually add to test_duecredit? (sounds like an easy GSoC starter?)

orbeckst replied:
I didnt' think we tested any of the other due citations. BUT: I was wrong! 

@edisj if you could add an entry in [testsuite/MDAnalysisTests/utils/test_duecredit.py](https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/utils/test_duecredit.py) then that would be the dot on the i.

(I don't know if we have really tested all duecredits. THat's something to check & easy enough)

edisj replied:
duecredit test is in

IAlibay commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -726,6 +801,642 @@ def __setstate__(self, state):
+    The H5MD format is very flexible with regards to units, as there is no
+    standard defined unit for the format. For this reason, :class:`H5MDWriter`
+    does not enforce any units. The units of the written trajectory can be set
+    explicitly with the keyword arguments `lengthunit`, `velocityunit`, and
+    `forceunit`. If units are not explicitly specified, they are set to the
```
```suggestion
    explicitly with the keyword arguments ``lengthunit``, ``velocityunit``, and
    ``forceunit``. If units are not explicitly specified, they are set to the
```

Do we want double ticks here? There's double ticks above for other parameters I _think_?

orbeckst replied:
I _think_ that when we refer to parameters in docstrings we use single ticks, i.e., the reST default role. If it's anything else code-like for which there's no reST role in sphinx then we use the "monospace" double-backticks.

(I didn't look at the code context here so I don't know if `lengthunit` and friends are parameters/kwargs of a method or no.)

edisj replied:
They're parameters of H5MDWriter's `__init__()`, so I _think_ they should get double back-ticks, or not?

orbeckst replied:
single back ticks

edisj replied:
I've been way too liberal with doubles... fixing now

IAlibay reviewed (approved):
Just the one comment from a quick scan though. I'll directly approve now since it's just a documentation style thing.

edisj commented:
@IAlibay @orbeckst just merged with current develop for good measure, but I think the final points regarding the duecredit test and the backticks are resolved

richardjgowers commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -70,9 +76,40 @@
+        for ts in u.trajectory:
+            W.write(u)
+
+To write an H5MD file with contiguous datasets, you must specifly the
```
Specifly 🪰

IAlibay replied:
```suggestion
To write an H5MD file with contiguous datasets, you must specify the
```
I'll just merge it in myself.

richardjgowers commented on package/MDAnalysis/coordinates/H5MD.py:
```diff
@@ -70,9 +76,40 @@
+            W.write(u)
+
+To write an H5MD file with contiguous datasets, you must specifly the
+number of frames to be written and set ``chunks=False``:
```
Maybe a quick one line hint as to why someone would want to do this

IAlibay replied:
There's a note for it on line 948, although not sure if it fully covers what you're asking @richardjgowers ?

richardjgowers commented on package/MDAnalysis/coordinates/base.py:
```diff
@@ -107,6 +107,12 @@
+
+      :class:`dict` that holds arbitrary per Timestep data
+
+      .. versionadded:: 0.11.0
```
2.0?

IAlibay replied:
This is a retrospective addition of docs for `data`, which would match release 0.11.0.

richardjgowers reviewed (commented):
Looks good, just a couple doc quibbles that can slide

IAlibay commented:
Seems to be everything, I'll merge once CI returns green - thanks for all the work @edisj 🎉

orbeckst commented:
Congratulations @edisj — PR with >200 comments and what turned out to be one of the more complicated formats.

Given that this could be considered, together with the SciPy paper, the capstone of your NSF REU experience, would you want to write a blog post for https://www.mdanalysis.org/blog/ where you summarize your experience and achievements?
